pyomo/contrib/mindtpy/single_tree.py

```diff
@@ -0,0 +1,257 @@
+                                  ignore_integrality=False):
+        """Copy variable values from one list to another.
+
+        Rounds to Binary/Integer if neccessary
+        Sets to zero for NonNegativeReals if neccessary
+        """
+        for v_from, v_to in zip(from_list, to_list):
+            if skip_stale and v_from.stale:
+                continue  # Skip stale variable values.
+            if skip_fixed and v_to.is_fixed():
+                continue  # Skip fixed variables.
+            try:
+                v_to.set_value(self.get_values(
+                    opt._pyomo_var_to_solver_var_map[v_from]))
+                if skip_stale:
+                    v_to.stale = False
+            except ValueError as err:
+                err_msg = getattr(err, 'message', str(err))
+                # get the value of current feasible solution
+                # self.get_value() is an inherent function from Cplex
+                var_val = self.get_values(
+                    opt._pyomo_var_to_solver_var_map[v_from])
+                rounded_val = int(round(var_val))
+                # Check to see if this is just a tolerance issue
+                if ignore_integrality \
+                    and ('is not in domain Binary' in err_msg
+                         or 'is not in domain Integers' in err_msg):
+                    v_to.value = self.get_values(
+                        opt._pyomo_var_to_solver_var_map[v_from])
+                elif 'is not in domain Binary' in err_msg and (
+                        fabs(var_val - 1) <= config.integer_tolerance or
+                        fabs(var_val) <= config.integer_tolerance):
+                    v_to.set_value(rounded_val)
+                # TODO What about PositiveIntegers etc?
+                elif 'is not in domain Integers' in err_msg and (
+                        fabs(var_val - rounded_val) <= config.integer_tolerance):
+                    v_to.set_value(rounded_val)
+                # Value is zero, but shows up as slightly less than zero.
+                elif 'is not in domain NonNegativeReals' in err_msg and (
+                        fabs(var_val) <= config.zero_tolerance):
+                    v_to.set_value(0)
+                else:
+                    raise
+
+    def add_lazy_oa_cuts(self, target_model, dual_values, solve_data, config, opt,
+                         linearize_active=True,
+                         linearize_violated=True,
+                         linearize_inactive=False,
+                         use_slack_var=False):
+        """Add oa_cuts through Cplex inherent function self.add()"""
+
+        for (constr, dual_value) in zip(target_model.MindtPy_utils.constraint_list,
+                                        dual_values):
+            if constr.body.polynomial_degree() in (0, 1):
+                continue
+
+            constr_vars = list(identify_variables(constr.body))
+            jacs = solve_data.jacobians
```

Comment:
Can we guarantee this is always available? If I remember correctly we need to calculate them first.

Reply:
Since we initialize the iteration loop by evaluating the Jacobian, this case should never happen.